### PR TITLE
Add support for sampling

### PR DESCRIPTION
**Pull request by bobbobbio**
I've been working on some Rust code of my own to create cpu profiles using the linux perf-event sampling API. I saw that you already had a perf-event project going. I tried my hand at adding in some sampling code. Hopefully the PR is welcome. Let me know what needs to be fixed.

### CODE REVIEW

**Comment by jimblandy** [comment]
Oh, this looks really interesting. I am in Germany this week for a Mozilla meeting, but I'll look this over early next week.

File: src/lib.rs [hunk 1]
```diff
@@ -479,6 +595,18 @@ impl<'a> Builder<'a> {
 
         Ok(Counter { file, id })
     }
+
+    /// Construct a [`SampleStream`].
+    ///
+    /// A freshly built `SampleStream` is disabled. To being reading records from the read, you
```
Comment (jimblandy) [inline src/lib.rs]: "to begin reading records from the stream"

File: src/sample.rs [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+use std::os::raw::c_void;
+
+/// Controls the various fields that are provided in [`PerfRecordSample`] when sampling.
+///
+/// Corresponds to the `sample_type` field `perf_event_attr` in the [`perf_event_open`][man] man
+/// page.
+///
+/// [man]: http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+///
+/// Not all possible values of this enum are currently included
```
Comment (jimblandy) [inline src/sample.rs]: missing period at end of sentence

File: src/lib.rs [hunk 3]
```diff
@@ -421,6 +443,83 @@ impl<'a> Builder<'a> {
         self
     }
 
+    /// When sampling, include the current instruction pointer.
+    pub fn sample_ip(mut self) -> Builder<'a> {
```
Comment (jimblandy) [inline src/lib.rs]: I think it would be better to consolidate the methods that just set a bit in `self.sample_type_set` into a single `sample` method that takes a `PerfSampleType` value as an argument. The documentation on each function explaining what it adds is great, but it belongs on the `PerfSampleType` variants themselves.

File: src/lib.rs [hunk 4]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
...
+            // XXX its hard to verify other stuff about the sample since we can't predict what
+            // the value are.
+            } else {
+                panic!();
+            }
+        }
+        DONE.store(true, Ordering::Relaxed);
+    });
+
+    while !DONE.load(Ordering::Relaxed) {
```
Comment (jimblandy) [inline src/lib.rs]: Add a comment:
```
/// This busy-wait loop creates activity for the sampling thread to observe.
```

File: src/lib.rs [hunk 5]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
...
+
+    sample_stream.enable()?;
+
+    static DONE: AtomicBool = AtomicBool::new(false);
+
+    let current_pid = unsafe { libc::getpid() };
+
+    // Sample on a different thread and create samples on the main thread until we get at least
+    // ten.
+    std::thread::spawn(move || {
```
Comment (jimblandy) [inline src/lib.rs]: You should keep the `JoinHandle` that `spawn` returns, and then do a `join()?` on it after waiting for `DONE`, so that if it panics, the main thread will panic too.

File: src/lib.rs [hunk 6]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
...
+        .sample_tid()
+        .sample_time()
+        .sample_cpu()
+        .sample_period()
+        .sample_callchain()
+        .sample_stream()?;
+
+    sample_stream.enable()?;
+
+    static DONE: AtomicBool = AtomicBool::new(false);
```
Comment (jimblandy) [inline src/lib.rs]: I would rather see an `Arc<AtomicBool>` here, and let the closure capture a clone, than use a `static`.

File: src/lib.rs [hunk 7]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
...
+    // Sample on a different thread and create samples on the main thread until we get at least
+    // ten.
+    std::thread::spawn(move || {
+        for _ in 0..10 {
+            if let Some(PerfRecord::Sample(sample)) = sample_stream.read(None).unwrap() {
+                // We should only get samples for the pid we asked for.
+                assert_eq!(sample.pid.unwrap(), current_pid);
+
+            // XXX its hard to verify other stuff about the sample since we can't predict what
+            // the value are.
```
Comment (jimblandy) [inline src/lib.rs]: `XXX` or `TODO` comments are usually material that belongs in a follow-up issue instead. I'd rather use an issue tracker than grep the source tree.

File: src/lib.rs [hunk 8]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
         .counter()
         .expect("Couldn't build default Counter");
 }
+
+// Use a pretty big buffer because we don't want to drop any entries
+const SAMPLE_BUFFER_SIZE: usize = 528384;
```
Comment (jimblandy) [inline src/lib.rs]: It would be a little less mysterious to write this as `(128 + 1) * 4096`. That way, if someone has read the docs, they'll be able to say, "Oh, 128 data pages plus the control page. Hmm, and they've hard-coded the page size..." :D 

If you want to not hard-code the size, the official way to get the page size on Linux is to call `sysconf(_SC_PAGESIZE)`, from the `libc` crate.

File: src/lib.rs [hunk 9]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
...
+    aux_tail: u64,
+    aux_offset: u64,
+    aux_size: u64,
+}
+
+/// A stream of samples being sent to us from the kernel. These samples represent instantaneous
+/// states or events concerning the process(es) being profiled.
+///
+/// Internally the samples are queued up in a ring-buffer. The kernel writes samples into the
+/// buffer, and [`SampleStream.read`] deques them. If the buffer it full, the kernel will overwrite
```
Comment (jimblandy) [inline src/lib.rs]: "dequeues"
"if the buffer is full"

File: src/lib.rs [hunk 10]
```diff
@@ -806,3 +934,241 @@ fn simple_build() {
...
+        // from that. It just makes things a bit trickier.
+
+        // wait for there to be data in the buffer, or the timeout.
+        if wait_for_readable_or_timeout(&self.file, timeout) {
+            return Ok(None);
+        }
+
+        // The kernel gives us records in a ring buffer. As the kernel adds records to the head, we
+        // are consuming from the tail. If the buffer is full, the kernel drops records.
+        let header: *mut PerfEventMmapPage = unsafe { std::mem::transmute(self.mapped_memory) };
```
Comment (jimblandy) [inline src/lib.rs]: This is just turning a `*mut c_void` into a `*mut PerfEventMmapPage`, right? Oddly, you can just use `as *mut PerfEventMmapPage` to do this, no `transmute` required, and hence no `unsafe` required.

**Review by jimblandy — COMMENTED** [review]
Thanks very much for this! I ran out of time to finish the review; here's what I've got so far. More tomorrow.

File: src/lib.rs (hunk 1 above)
Comment (bobbobbio) [reply src/lib.rs]: fixed

File: src/sample.rs (hunk 2 above)
Comment (bobbobbio) [reply src/sample.rs]: fixed

File: src/lib.rs (hunk 3 above)
Comment (bobbobbio) [reply src/lib.rs]: applied

File: src/lib.rs (hunk 4 above)
Comment (bobbobbio) [reply src/lib.rs]: added. Didn't make it a doc-comment, there is a warning about unused doc-comments

File: src/lib.rs (hunk 5 above)
Comment (bobbobbio) [reply src/lib.rs]: done

File: src/lib.rs (hunk 6 above)
Comment (bobbobbio) [reply src/lib.rs]: done

File: src/lib.rs (hunk 7 above)
Comment (bobbobbio) [reply src/lib.rs]: I removed it. I'm not sure it needs an issue really.

File: src/lib.rs (hunk 8 above)
Comment (bobbobbio) [reply src/lib.rs]: fixed

File: src/lib.rs (hunk 9 above)
Comment (bobbobbio) [reply src/lib.rs]: fixed

File: src/lib.rs (hunk 10 above)
Comment (bobbobbio) [reply src/lib.rs]: fixed

**Comment by jimblandy** [comment]
Fantastic. I am in a crunch right now with work and book, but I will try to look this over as soon as I can.

File: src/lib.rs [hunk 11]
```diff
@@ -795,5 +886,252 @@ where F: FnOnce() -> R,
...
+    fn new(attrs: sys::bindings::perf_event_attr, file: File) -> std::io::Result<Self> {
+        let mapped_memory = check_syscall(|| unsafe {
+            mmap(
+                std::ptr::null_mut(),
+                SAMPLE_BUFFER_PAGES * page_size(),
+                PROT_READ | PROT_WRITE,
+                MAP_SHARED,
+                file.as_raw_fd(),
+                0,
+            ) as isize
```
Comment (eddyb) [inline src/lib.rs:1236]: I've personally used the `memmap` crate for convenience, I wonder how much sense that would make here?

File: src/lib.rs [hunk 12]
```diff
@@ -795,5 +886,252 @@ where F: FnOnce() -> R,
...
+    time_shift: u16,
+    time_mult: u32,
+    time_offset: u64,
+    __reserved: [u64; 120], /* Pad to 1 k */
+
+    /// head in the data section
+    data_head: AtomicU64,
+
+    /// user-space written tail
+    data_tail: AtomicU64,
```
Comment (eddyb) [inline src/lib.rs:1197]: I wonder if casting `*const u64` pointers to `&AtomicU64` (or even having a `PerfEventDataBuffer` struct with the `data_*` fields) would be preferable to copying the `struct`.

Or maybe there's a way to have `bindgen` use different types for some fields in the `-sys` crate?

(I'm interested in this because it's adjacent to the `rdpmc` usage, although all examples I've seen for that just compare the value of `lock` before/after `rdpmc`, and use barriers, without any otherwise-synchronized loads)

**Comment by jimblandy** [comment]
@eddyb If you're interested, I would love help getting this reviewed and merged. I'm overwhelmed at the moment, but if you need this and can give the API and code a look-through, I'd be happy to add you to the repo as a collaborator.

**Comment by eddyb** [comment]
@jimblandy Thanks for the offer! Although I've been overwhelmed as well, it just happens that I'm currently experimenting with fine-grained userspace instruction-counting for [`measureme`](https://github.com/rust-lang/measureme) (which powers `rustc -Z self-profile`), using `rdpmc`.

Thanks to `perf-event-open-sys`, I don't need too much to get started, and if I can make a nice API for it I'll try to upstream whichever design I end up with (I want to find the lowest-overhead method I can, as some of the profiled functions are *tiny*).

**Comment by jimblandy** [comment]
> Thanks to perf-event-open-sys, I don't need too much to get started, and if I can make a nice API for it I'll try to upstream whichever design I end up with (I want to find the lowest-overhead method I can, as some of the profiled functions are tiny).

I was hoping this crate's API would be nice! :D

The `enable` and `disable` calls should compile to nothing more than an `ioctl` and some error checking, which seems like the best you can do with the `perf_event_open` API. It uses `File` internally, but an unbuffered `File` is nothing more than the file descriptor with the `Drop` impl you'll need anyway. The only way to streamline things that I can see would be to have `Builder` construct the `perf_event_attr` structure as it goes along.

In any case, I'll look forward to what you come up with!

**Comment by eddyb** [comment]
@jimblandy To expand a bit: I plan to use `rdpmc` to avoid incurring the cost of a syscall, and the remaining costs I want to golf have to do with the few memory reads from the perf mmap page, around the `rdpmc` instruction (search for "rdpmc" in the `perf_event_open` manpage for an example).

Oh and I guess because it's inline assembly, this low-overhead approach would be nightly-only for a while.

**Comment by blt** [comment]
@jimblandy this is something I'd be very interested in helping getting merged, if you'd be interested in having it resurrected.

**Comment by jimblandy** [comment]
Thanks! My first concern is that I haven't actually tried out the proposed API yet. Have you used this patch in your own work? What's it like?

**Comment by blt** [comment]
Oh yeah, sure thing. For context I'm in-progress on building a causal profiler for [vector](https://github.com/vectordotdev/vector/) so that's where my interest in samples comes from. Here's the example program I built to satisfy myself that the patch was at least nominally functional, after having read it over:

```rust
use perf_event::events::Hardware;
use perf_event::sample::{PerfRecord, PerfSampleType};
use perf_event::Builder;
use std::num::NonZeroU64;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc,
};

fn collatz(n: NonZeroU64) -> u64 {
    let mut cur = n.get();
    let mut iters = 0;
    while cur != 1 {
        iters += 1;
        if cur % 2 == 0 {
            cur = cur / 2;
        } else {
            cur = (3 * cur) + 1;
        }
    }
    iters
}

fn main() -> std::io::Result<()> {
    let sampler = Builder::new()
        .kind(Hardware::CPU_CYCLES)
        .sample_frequency(1000)
        .sample(PerfSampleType::ADDR)
        .sample(PerfSampleType::CALLCHAIN)
        .sample(PerfSampleType::CPU)
        .sample(PerfSampleType::IP)
        .sample(PerfSampleType::PERIOD)
        .sample(PerfSampleType::STREAM_ID)
        .sample(PerfSampleType::TID)
        .sample(PerfSampleType::TIME)
        .sample_stream()?;

    sampler.enable()?;

    let do_sample = Arc::new(AtomicBool::new(true));

    // This is the area of the program that is doing the sampling. It runs until
    // signaled to stop from the main thread. The sampler is dropped to disable
    // it.
    let sampler_do_sample = Arc::clone(&do_sample);
    let handle = std::thread::spawn(move || {
        let do_sample = sampler_do_sample;
        while do_sample.load(Ordering::Relaxed) {
            if let Some(PerfRecord::Sample(sample)) = sampler.read(None).unwrap() {
                println!("{:?}", sample);
            } else {
                panic!();
            }
        }
        drop(sampler);
    });

    // This is the area of the program that is being sampled.
    for _ in 0..100_000 {
        for i in 1..100 {
            collatz(NonZeroU64::new(i).unwrap());
        }
    }
    do_sample.store(false, Ordering::Relaxed);

    handle.join().unwrap();

    Ok(())
}
```

I'm in-progress on reconstructing stack traces from samples but what I have is too much of a mess to be convincing one way or another. I'm happy to keep using this patch and report back on odd problems I hit, more elaborate success, if you'd like. 

About the only awkward thing with the API as proposed is the lack of an explicit disable on the streamer. It flips off when dropped. Otherwise, absent any bugs I find in the next few days, it works well.

**Comment by bobbobbio** [comment]
I'd be happy to make changes to this to get it merged.

I was interested in implementing some Rust on Linux tailored CPU profiling tools and was hoping to help create safe bindings for everything Linux offers, but I mostly just gave up waiting for someone else to be interested enough to help get stuff merged here or whatever

**Comment by blt** [comment]
> I'd be happy to make changes to this to get it merged.

Rad, thank you!

> I was interested in implementing some Rust on Linux tailored CPU profiling tools and was hoping to help create safe bindings for everything Linux offers, but I mostly just gave up waiting for someone else to be interested enough to help get stuff merged here or whatever

Totally makes sense. It's an... elaborate programming target. FWIW the current implementation works for my purposes and I'd be glad to see it landed, fiddle with the performance notes in the PR as my work progresses.

File: src/sample.rs [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+}
+
+/// This record is a sample of the current state of some process.
+///
+/// Corresponds to the anonymous struct under `PERF_RECORD_SAMPLE` in [`perf_event_open`][man] man
+/// page.
+///
+/// [man]: http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct PerfRecordSample {
```
Comment (mstange) [inline src/sample.rs:142]: I'm currently working on [a `linux-perf-event-reader` crate](https://docs.rs/linux-perf-event-reader/) which has some complete code for parsing these records. It also supports "wrap-around" reading without copies, via [`RawData::Split`](https://docs.rs/linux-perf-event-reader/0.4.0/linux_perf_event_reader/enum.RawData.html).

Comment (jimblandy) [reply src/sample.rs:142]: Oh, that's great!

One of the reasons I haven't merged this PR is that I am concerned about testability. I don't want to have long sections of detailed, technical code that I have no way to ensure coverage for. But this weekend I finally got around to adding an interface for mocking the underlying system calls, so we can write nice deterministic tests that cover all the edge cases. So hopefully we'll be able to make some progress here.

**Comment by Phantomical** [comment]
@jimblandy @bobbobbio It doesn't look like there's any action here so is it alright if I revive this PR and try and bring it to a workable state?

**Comment by jimblandy** [comment]
Sure! We're looking for an API that is both safe, and close enough to what's described in the open_perf_event man page that we don't have to redocument the entire thing for it to be usable.

The thing I was most worried about here was that it would add a lot of complexity - and I want that to be covered by tests. The "hooks" feature was supposed to give me a way to write tests that could supply arbitrary data and force the use of every code path, without having to get the real kernel involved - because there's no way to ensure that's going to give you exactly the data you need to properly test the code.

But, nice clean code is always welcome, even if going that far with testing isn't your cup of tea. So yes, it'd be great to have something here.

**Comment by jimblandy** [comment]
This work has been taken up in some newer PRs. Thanks very much to @bobbobbio for this initial work!
